policy/policy.go

```diff
@@ -44,85 +45,198 @@ func (policy *PolicySvc) Routes() common.Routes {
+// by looking it up in the appropriate service.
+func (policy *PolicySvc) augmentEndpoint(endpoint *common.Endpoint) error {
+	tenantSvcUrl, err := policy.client.GetServiceUrl("tenant")
 	if err != nil {
-		return nil, err
+		return err
+	}
+
+	// TODO this will have to be changed once we implement
+	// https://paninetworks.kanbanize.com/ctrl_board/3/cards/319/details
+	var tenantIDToUse string
+	ten := &tenant.Tenant{}
+	if endpoint.TenantID != 0 {
+		tenantIDToUse = string(endpoint.TenantID)
+	} else if endpoint.TenantExternalID != "" {
+		tenantIDToUse = endpoint.TenantExternalID
+	}
+	if tenantIDToUse != "" {
+		tenantsUrl := fmt.Sprintf("%s/tenants/%s", tenantSvcUrl, tenantIDToUse)
+		err = policy.client.Get(tenantsUrl, &ten)
+		if err != nil {
+			return err
+		}
+		endpoint.TenantNetworkID = ten.Seq
+	}
+
+	var segmentIDToUse string
+	if endpoint.SegmentID != 0 {
+		segmentIDToUse = string(endpoint.SegmentID)
+	} else if endpoint.SegmentExternalID != "" {
+		segmentIDToUse = endpoint.SegmentExternalID
 	}
-	// Get host info from topology service
-	topoUrl, err := client.GetServiceUrl("topology")
+	if segmentIDToUse != "" {
+		tenantsUrl := fmt.Sprintf("%s/tenants/%s/segment/%s", tenantSvcUrl, ten.ID, segmentIDToUse)
+		segment := &tenant.Segment{}
+		err = policy.client.Get(tenantsUrl, &segment)
+		if err != nil {
+			return err
+		}
+		endpoint.SegmentNetworkID = segment.Seq
+	}
+	return nil
+}
+
+// augmentPolicy augments the provided policy with information gotten from
+// various services.
+func (policy *PolicySvc) augmentPolicy(policyDoc *common.Policy) error {
+	// Get info from topology service
+	topoUrl, err := policy.client.GetServiceUrl("topology")
 	if err != nil {
-		return nil, err
+		return err
 	}
 
+	// Query topology for data center information
+	// TODO move this to root
 	index := common.IndexResponse{}
-	err = client.Get(topoUrl, &index)
+	err = policy.client.Get(topoUrl, &index)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	tenantSvcUrl, err := client.GetServiceUrl("tenant")
+	dcURL := index.Links.FindByRel("datacenter")
+	dc := common.Datacenter{}
+	err = policy.client.Get(dcURL, &dc)
 	if err != nil {
-		return nil, err
+		return err
+	}
+	log.Printf("Policy server received datacenter information from topology service: %v\n", dc)
+	policyDoc.Datacenter = dc
+
+	for _, endpoint := range policyDoc.AppliedTo {
+		err = policy.augmentEndpoint(&endpoint)
+		if err != nil {
+			return err
+		}
 	}
 
-//	for _, endpoint := range policyDoc.AppliedTo {
-//	}
-	tenantsUrl := fmt.Sprintf("%s/tenants", tenantSvcUrl)
-	var tenants []tenant.Tenant
-	err = client.Get(tenantsUrl, &tenants)
+	for _, endpoint := range policyDoc.Peers {
+		err = policy.augmentEndpoint(&endpoint)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// distributePolicy distributes policy to all agents.
+// TODO how should error handling work here really?
```

Comment:
I think that a host on which a policy can't be applied should be taken out of rotation!

Is there a way to signal back to the environment (OpenStack or Kubernetes) that a host is in trouble?